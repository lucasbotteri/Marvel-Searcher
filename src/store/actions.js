import {
  SET_LOADING_CHARACTERS_LIST,
  SET_CHARACTERS_LIST,
  SET_CHARACTER_SELECTED,
  HIDE_MODAL,
  SET_COMICS,
  SET_LOADING_COMICS,
  SET_COMIC,
  SET_LOADING_COMIC,
  SAVE_FAV,
  SHOW_FAVS,
  REMOVE_FAV,
  HIDE_FAVS,
} from "./constants";
import {
  getCharacterByName,
  getRandomCharacter,
  getCharacterComics,
  getComic,
} from "../service";

export const setLoadingCharacters = () => ({
  type: SET_LOADING_CHARACTERS_LIST,
});

export const setCharacters = (characters) => ({
  type: SET_CHARACTERS_LIST,
  payload: characters,
});

export const setCharacterSelected = (id, name) => ({
  type: SET_CHARACTER_SELECTED,
  payload: { id, name },
});

export const hideModal = () => ({
  type: HIDE_MODAL,
});

export const setComics = (comics) => ({
  type: SET_COMICS,
  payload: comics,
});

export const setLoadingComics = () => ({
  type: SET_LOADING_COMICS,
});

export const setComic = (comics) => ({
  type: SET_COMIC,
  payload: comics,
});

export const setLoadingComic = () => ({
  type: SET_LOADING_COMIC,
});

export const saveFav = (character) => ({
  type: SAVE_FAV,
  payload: character,
});

export const showFavs = () => ({
  type: SHOW_FAVS,
});

export const hideFavs = () => ({
  type: HIDE_FAVS,
});

export const removeFav = (id) => ({
  type: REMOVE_FAV,
  payload: { id },
});

export const searchCharacter = async (dispatch, characterName) => {
  dispatch(setLoadingCharacters());
  const characters = await getCharacterByName(characterName);
  dispatch(setCharacters(characters));
};

export const searchRandomCharacter = async (dispatch) => {
  dispatch(setLoadingCharacters());
  const characters = await getRandomCharacter();
  dispatch(setCharacters(characters));
};

export const searchCharacterComics = async (
  dispatch,
  characterId,
  comicFilter
) => {
  dispatch(setLoadingComics());
  const comics = await getCharacterComics(characterId, comicFilter);
  dispatch(setComics(comics));
};

export const searchComic = async (dispatch, comicId) => {
  dispatch(setLoadingComic());
  const comic = await getComic(comicId);
  dispatch(setComic(comic));
};
