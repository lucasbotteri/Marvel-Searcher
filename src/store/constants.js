export const LOADING_CHARACTERS_LIST = "LOADING_CHARACTERS_LIST";
export const SET_CHARACTERS_LIST = "SET_CHARACTERS_LIST";
export const SET_CHARACTER_SELECTED = "SET_CHARACTER_SELECTED";
export const HIDE_MODAL = "HIDE_MODAL";
export const SET_COMICS = "SET_COMICS";
export const SET_LOADING_COMICS = "SET_LOADING_COMICS";
