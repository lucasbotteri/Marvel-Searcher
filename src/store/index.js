export { default as StoreProvider, StoreContext } from "./StoreProvider";
