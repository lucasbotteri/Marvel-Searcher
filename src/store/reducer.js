import {
  LOADING_CHARACTERS_LIST,
  SET_CHARACTERS_LIST,
  SET_CHARACTER_SELECTED,
  HIDE_MODAL,
} from "./constants";

const reducer = (state, action) => {
  switch (action.type) {
    case LOADING_CHARACTERS_LIST:
      return { ...state, characters: [], isCharacterListLoading: true };
    case SET_CHARACTERS_LIST:
      return {
        ...state,
        characters: action.payload,
        isCharacterListLoading: false,
        isShowingModal: false,
      };
    case SET_CHARACTER_SELECTED:
      return {
        ...state,
        characterSelected: action.payload,
        isShowingModal: true,
      };
    case HIDE_MODAL:
      return {
        ...state,
        isShowingModal: false,
      };
    default:
      throw new Error("Invalid Action");
  }
};

export default reducer;
