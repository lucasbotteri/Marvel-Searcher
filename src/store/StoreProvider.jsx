import React, { useReducer } from "react";
import PropTypes from "prop-types";
import reducer from "./reducer";
import * as actions from "./actions";

const INITAL_STATE = {
  characters: [],
  loadingCharacters: false,
  fetchedCharactersSuccessful: true,
};

export const StoreContext = React.createContext();

const StoreProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, INITAL_STATE);
  const value = {
    ...state,
    fetchCharacters: (data) => actions.fetchCharacters(dispatch, data),
  };
  return (
    <StoreContext.Provider value={value}>{children}</StoreContext.Provider>
  );
};

StoreProvider.propTypes = {
  children: PropTypes.element.isRequired,
};

export default StoreProvider;
