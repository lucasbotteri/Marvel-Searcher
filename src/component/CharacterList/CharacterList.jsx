import React, { useContext } from "react";
import { StoreContext } from "../../store";
import styled from "styled-components";
import CharacterCard from "../CharacterCard";

const CharacterListWrapper = styled.ul`
  box-sizing: border-box;
  display: grid;
  grid-gap: 33px;
  grid-template-columns: repeat(auto-fit, minmax(256px, 1fr));
  justify-items: center;
  padding: 40px 40px;
  width: 100%;
`;

const CharacterListItem = styled.li`
  list-style: none;
`;

const CharacterList = () => {
  const { characters, selectCharacter } = useContext(StoreContext);

  return (
    <CharacterListWrapper>
      {characters.map((c) => (
        <CharacterListItem>
          <CharacterCard
            key={c.id}
            character={c}
            onClick={() => {
              selectCharacter(c.id, c.name);
            }}
          />
        </CharacterListItem>
      ))}
    </CharacterListWrapper>
  );
};

export default CharacterList;
