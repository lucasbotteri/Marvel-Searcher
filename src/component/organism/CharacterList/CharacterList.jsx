import React, { useContext } from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import { useQuery } from "../../../hook";
import { StoreContext } from "../../../store";
import CharacterCard from "../../molecule/CharacterCard";

const CharacterListWrapper = styled.ul`
  box-sizing: border-box;
  display: grid;
  grid-gap: 33px;
  grid-template-columns: repeat(auto-fit, minmax(256px, 1fr));
  justify-items: center;
  padding: 20px;
  margin: 0;
  width: 100%;
`;

const CharacterListItem = styled.li`
  list-style: none;
`;

const CharacterList = ({ characters }) => {
  const {
    selectCharacter,
    searchCharacterComics,
    saveFav,
    removeFav,
    favs,
  } = useContext(StoreContext);
  const { comic: comicFilter } = useQuery();

  return (
    <CharacterListWrapper>
      {characters.map((c) => (
        <CharacterListItem key={c.id}>
          <CharacterCard
            character={c}
            onClick={() => {
              selectCharacter(c.id, c.name);
              searchCharacterComics(c.id, comicFilter);
            }}
            isFaved={favs && favs.some((f) => f.id === c.id)}
            onFaved={() => saveFav(c)}
            onUnfaved={() => removeFav(c.id)}
          />
        </CharacterListItem>
      ))}
    </CharacterListWrapper>
  );
};

CharacterList.propTypes = {
  characters: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string,
      thumbnail: PropTypes.string,
    })
  ).isRequired,
};

export default CharacterList;
