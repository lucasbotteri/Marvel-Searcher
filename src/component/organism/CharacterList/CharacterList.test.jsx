import React from "react";
import { mount } from "enzyme";
import { MemoryRouter } from "react-router";
import CharacterList from "./CharacterList";
import { StoreContext } from "../../../store";

const characters = [
  {
    id: 1014997,
    name: "Deathstrike (Ultimate)",
    thumbnail: "someurl.com",
  },
];

const contextStore = {
  selectCharacter: () => {},
  searchCharacterComics: () => {},
  saveFav: () => {},
  removeFav: () => {},
  favs: null,
};

const mountCharacterList = () => {
  return mount(
    <MemoryRouter initialEntries={["/"]} keyLength={0}>
      <StoreContext.Provider value={contextStore}>
        <CharacterList characters={characters} />
      </StoreContext.Provider>
    </MemoryRouter>
  );
};

describe("CharacterList", () => {
  it("Should render without crashing", () => {
    const wrapper = mountCharacterList();
    expect(wrapper).toBeTruthy();
  });

  it("Should match snapshot", () => {
    const wrapper = mountCharacterList();
    expect(wrapper).toMatchSnapshot();
  });
});
