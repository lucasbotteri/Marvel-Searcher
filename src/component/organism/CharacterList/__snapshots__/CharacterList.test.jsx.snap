// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterList Should match snapshot 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      "/",
    ]
  }
  keyLength={0}
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <CharacterList
      characters={
        Array [
          Object {
            "id": 1014997,
            "name": "Deathstrike (Ultimate)",
            "thumbnail": "someurl.com",
          },
        ]
      }
    >
      <styled.ul>
        <ul
          className="sc-AxgMl bVLvwi"
        >
          <styled.li
            key="1014997"
          >
            <li
              className="sc-AxheI kdRVYU"
            >
              <CharacterCard
                character={
                  Object {
                    "id": 1014997,
                    "name": "Deathstrike (Ultimate)",
                    "thumbnail": "someurl.com",
                  }
                }
                isFaved={null}
                onClick={[Function]}
                onFaved={[Function]}
                onUnfaved={[Function]}
              >
                <styled.section
                  alt="Deathstrike (Ultimate)"
                  onClick={[Function]}
                >
                  <section
                    alt="Deathstrike (Ultimate)"
                    className="sc-AxirZ bCKUBB"
                    onClick={[Function]}
                  >
                    <styled.img
                      src="someurl.com"
                    >
                      <img
                        className="sc-AxiKw jTQMHy"
                        src="someurl.com"
                      />
                    </styled.img>
                    <Styled(Star)
                      isSolid={null}
                      onClick={[Function]}
                    >
                      <Star
                        className="sc-AxhUy iYWvJI"
                        isSolid={null}
                        onClick={[Function]}
                      >
                        <Styled(FontAwesomeIcon)
                          className="sc-AxhUy iYWvJI"
                          icon={
                            Object {
                              "icon": Array [
                                576,
                                512,
                                Array [],
                                "f005",
                                "M528.1 171.5L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6zM388.6 312.3l23.7 138.4L288 385.4l-124.3 65.3 23.7-138.4-100.6-98 139-20.2 62.2-126 62.2 126 139 20.2-100.6 98z",
                              ],
                              "iconName": "star",
                              "prefix": "far",
                            }
                          }
                          onClick={[Function]}
                        >
                          <FontAwesomeIcon
                            border={false}
                            className="sc-AxjAm jumYnU sc-AxhUy iYWvJI"
                            fixedWidth={false}
                            flip={null}
                            icon={
                              Object {
                                "icon": Array [
                                  576,
                                  512,
                                  Array [],
                                  "f005",
                                  "M528.1 171.5L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6zM388.6 312.3l23.7 138.4L288 385.4l-124.3 65.3 23.7-138.4-100.6-98 139-20.2 62.2-126 62.2 126 139 20.2-100.6 98z",
                                ],
                                "iconName": "star",
                                "prefix": "far",
                              }
                            }
                            inverse={false}
                            listItem={false}
                            mask={null}
                            onClick={[Function]}
                            pull={null}
                            pulse={false}
                            rotation={null}
                            size={null}
                            spin={false}
                            swapOpacity={false}
                            symbol={false}
                            title=""
                            transform={null}
                          >
                            <svg
                              aria-hidden="true"
                              className="svg-inline--fa fa-star fa-w-18 sc-AxjAm jumYnU sc-AxhUy iYWvJI"
                              data-icon="star"
                              data-prefix="far"
                              focusable="false"
                              onClick={[Function]}
                              role="img"
                              style={Object {}}
                              viewBox="0 0 576 512"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                d="M528.1 171.5L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6zM388.6 312.3l23.7 138.4L288 385.4l-124.3 65.3 23.7-138.4-100.6-98 139-20.2 62.2-126 62.2 126 139 20.2-100.6 98z"
                                fill="currentColor"
                                style={Object {}}
                              />
                            </svg>
                          </FontAwesomeIcon>
                        </Styled(FontAwesomeIcon)>
                      </Star>
                    </Styled(Star)>
                    <styled.h2>
                      <h2
                        className="sc-AxhCb lbMrWD"
                      >
                         
                        Deathstrike (Ultimate)
                         
                      </h2>
                    </styled.h2>
                  </section>
                </styled.section>
              </CharacterCard>
            </li>
          </styled.li>
        </ul>
      </styled.ul>
    </CharacterList>
  </Router>
</MemoryRouter>
`;
